File: crates/domain/src/library/library.rs
use jiff::Timestamp;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct LibraryId(pub String);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum LibraryKind {
    Movie,
    Tv,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum WatcherStrategy {
    Local,
    Polling,
    Scheduled,
    Manual,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum LibraryOrigin {
    Local,
    External,
}

#[derive(Debug, Clone)]
pub struct Library {
    pub id: LibraryId,
    pub name: String,
    pub kind: LibraryKind,
    pub origin: LibraryOrigin,
    pub roots: Vec<String>,
    pub watcher: WatcherStrategy,
    pub scan_schedule: Option<String>,
    pub metadata_sources: Vec<String>,
    pub sort_articles: Vec<String>,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}

File: clients/flutter/lib/model/library/library.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'library.freezed.dart';
part 'library.g.dart';

enum LibraryKind {
  @JsonValue('movie')
  movie,
  @JsonValue('tv')
  tv,
}

enum LibraryOrigin {
  @JsonValue('local')
  local,
  @JsonValue('external')
  external,
}

enum WatcherStrategy {
  @JsonValue('local')
  local,
  @JsonValue('polling')
  polling,
  @JsonValue('scheduled')
  scheduled,
  @JsonValue('manual')
  manual,
}

@freezed
abstract class Library with _$Library {
  const factory Library({
    required String id,
    required String name,
    required LibraryKind kind,
    @Default(LibraryOrigin.local) LibraryOrigin origin,
    @Default(<String>[]) List<String> roots,
    required WatcherStrategy watcher,
    String? scanSchedule,
    @Default(<String>[]) List<String> metadataSources,
    @Default(<String>[]) List<String> sortArticles,
    required String createdAt,
    required String updatedAt,
  }) = _Library;

  factory Library.fromJson(Map<String, dynamic> json) =>
      _$LibraryFromJson(json);
}
